### PR TITLE
Initial UI for Web apps

### PR DESCRIPTION
**Pull request by felipeerias**
This PR is the initial implementation of an UI for Web apps.

The user can add a new Web app by clicking on the `"Save Web app…"` button on the top bottom menu on the browser. This opens a dialog with the Web app's details. There is also a button in the URL bar with the same functionality.

Saved Web apps are displayed in a new tab in the Library.

### CODE REVIEW

**Comment by felipeerias** [comment]
This PR is finally ready for review. Please try it out and let me know any issues and suggestions.

In terms of the UI, I have a doubt regarding where we should put the "Save Web app…" button. For now, I have put it in two places: in the URL bar and in the menu in the bottom right corner:

* Placing this button in the URL bar has the issue that it might be confusing for users, because it looks similar to a bookmark.
* Placing it in the menu is not discoverable, because you need to open the menu just to check if the current website offers a manifest.

For now, saved Web apps work pretty much like regular bookmarks.

File: app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java [hunk 1]
```diff
@@ -87,38 +67,53 @@ public String getId() {
...
+        return mManifest.getStartUrl();
     }
 
     public int getThemeColor() {
-        return mThemeColor;
+        if (mManifest.getThemeColor() != null) {
+            return mManifest.getThemeColor();
+        } else {
+            return Color.TRANSPARENT;
+        }
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java:90]: I like the ternary operator for these cases
```
return mManifest.getThemeColor() != null ? mManifest.getThemeColor() : Color.TRANSPARENT
```
you can also cache the result of `getThemeColor()` although it isn't a big deal

File: app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java [hunk 2]
```diff
@@ -87,38 +67,53 @@ public String getId() {
...
     }
 
     public int getBackgroundColor() {
-        return mBackgroundColor;
+        if (mManifest.getBackgroundColor() != null) {
+            return mManifest.getBackgroundColor();
+        } else {
+            return Color.TRANSPARENT;
+        }
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java:98]: Ditto about ternary operator

File: app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java [hunk 3]
```diff
@@ -87,38 +67,53 @@ public String getId() {
...
+
+    // Returns the list of icons in a format suitable to be used with BrowserIcons.
+    public List<IconRequest.Resource> getIconResources() {
+        return mManifest.getIcons().stream().map(icon ->
+                new IconRequest.Resource(
+                        icon.getSrc(),
+                        IconRequest.Resource.Type.MANIFEST_ICON,
+                        icon.getSizes(),
+                        icon.getType(),
+                        icon.getPurpose().contains(WebAppManifest.Icon.Purpose.MASKABLE) || icon.getPurpose().contains(WebAppManifest.Icon.Purpose.ANY)
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java:109]: Is `getPurpouse` guaranteed to return non-null?

File: app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java [hunk 4]
```diff
@@ -130,19 +125,19 @@ public boolean equals(Object o) {
...
-                mScope.equals(webApp.mScope) &&
-                mStartUrl.equals(webApp.mStartUrl) &&
-                mThemeColor == webApp.mThemeColor &&
-                mBackgroundColor == webApp.mBackgroundColor;
+                getName().equals(webApp.getName()) &&
+                getShortName().equals(webApp.getShortName()) &&
+                getScope().equals(webApp.getScope()) &&
+                getStartUrl().equals(webApp.getStartUrl()) &&
+                getThemeColor() == webApp.getThemeColor() &&
+                getBackgroundColor() == webApp.getBackgroundColor();
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java:133]: Doesn't WebApp implement the `equals()` method?

File: app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java [hunk 5]
```diff
@@ -25,18 +30,7 @@ public class WebApp {
...
-    private String mShortName;
-    @NonNull
-    private String mScope;
-    @NonNull
-    private String mStartUrl;
-
-    private int mThemeColor;
-    private int mBackgroundColor;
-
+    private WebAppManifest mManifest;
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java:33]: Nice!

**Review by svillar — CHANGES_REQUESTED** [review]
Looks great I just have some nits

File: app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java (hunk 1 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java:90]: Done.

File: app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java (hunk 2 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java:98]: Done.

File: app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java (hunk 4 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java:133]: You're right, `WebAppManifest` is a "data class" and those get `equals()` for free in Kotlin: https://kotlinlang.org/docs/data-classes.html

File: app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java (hunk 5 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java:33]: :slightly_smiling_face:

**Review by svillar — APPROVED** [review]
Cool stuff

**Comment by svillar** [comment]
This is the initial implementation of an UI for Web apps.

The user can add a new Web app by clicking on the "Save Web app…" button on the top bottom menu on the browser. This opens a dialog with the Web app's details. There is also a button in the URL bar with the same functionality.

Saved Web apps are displayed in a new tab in the Library.

**Pull request merged**